napi-inl.h

```diff
@@ -975,41 +979,72 @@ inline Function::CallbackData::CallbackData(FunctionCallback cb, void* data)
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-// Buffer class
+// Buffer<T> class
 ////////////////////////////////////////////////////////////////////////////////
 
-inline Buffer Buffer::New(Napi::Env env, char* data, size_t size) {
+template <typename T>
+inline Buffer<T> Buffer<T>::New(Napi::Env env, size_t length) {
   napi_value value;
-  napi_status status = napi_buffer_new(env, data, size, &value);
+  char* data;
+  napi_status status = napi_create_buffer(env, length * sizeof(T), &data, &value);
   if (status != napi_ok) throw Error::New(env);
-  return Buffer(env, value);
+  return Buffer(env, value, length, data);
+}
+
+template <typename T>
+inline Buffer<T> Buffer<T>::New(
+    Napi::Env env, T* data, size_t length, napi_finalize finalizeCallback) {
+  napi_value value;
+  napi_status status = napi_create_external_buffer(
+    env, length * sizeof(T), reinterpret_cast<char*>(data), finalizeCallback, &value);
+  if (status != napi_ok) throw Error::New(env);
+  return Buffer(env, value, length, data);
 }
 
-inline Buffer Buffer::Copy(Napi::Env env, const char* data, size_t size) {
+template <typename T>
+inline Buffer<T> Buffer<T>::Copy(Napi::Env env, const T* data, size_t length) {
   napi_value value;
-  napi_status status = napi_buffer_copy(env, data, size, &value);
+  napi_status status = napi_create_buffer_copy(
+    env, reinterpret_cast<const char*>(data), length * sizeof(T), &value);
   if (status != napi_ok) throw Error::New(env);
   return Buffer(env, value);
 }
 
-inline Buffer::Buffer() : Object() {
+template <typename T>
+inline Buffer<T>::Buffer() : Object() {
 }
 
-inline Buffer::Buffer(napi_env env, napi_value value) : Object(env, value) {
+template <typename T>
+inline Buffer<T>::Buffer(napi_env env, napi_value value) : Object(env, value) {
 }
 
-inline size_t Buffer::Length() const {
-  size_t result;
-  napi_status status = napi_buffer_length(_env, _value, &result);
-  if (status != napi_ok) throw Error::New(Env());
-  return result;
+template <typename T>
+inline Buffer<T>::Buffer(napi_env env, napi_value value, size_t length, T* data)
+  : Object(env, value), _length(length), _data(data) {
 }
 
-inline char* Buffer::Data() const {
-  char* data;
-  napi_status status = napi_buffer_data(_env, _value, &data);
-  if (status != napi_ok) throw Error::New(Env());
-  return data;
+template <typename T>
+inline size_t Buffer<T>::Length() const {
+  if (_data == nullptr) {
```

Comment:
This same initialization code is duplicated in Buffer<T>::Length and Buffer<T>::Data. Would it make sense to put it into a helper like a private method Buffer<T>::EnsureCachedInfo or something to make the intent more clear and avoid having to change two places in the future? Also, might want to copy/paste this block again for another method later and inlining a small helper should translate to exactly the same code.